prow/tide/tide.go

```diff
@@ -854,8 +906,23 @@ func (c *Controller) takeAction(sp subpool, batchPending, successes, pendings, n
 func (c *Controller) syncSubpool(sp subpool) (Pool, error) {
 	sp.log.Infof("Syncing subpool: %d PRs, %d PJs.", len(sp.prs), len(sp.pjs))
 	var presubmits []string
+	bp := branchProtection{
+		config: c.ca.Config().Tide.BranchProtection.GetBranchProtection(sp.org, sp.repo, sp.branch),
+	}
+	if bp.config.UseRequiredChecksOnly {
+		requiredChecks, err := c.ghc.GetRequiredChecks(sp.org, sp.repo, sp.branch)
```

Comment:
Why should we fetch this twice (in `takeAction` and here) for the same controller?

Reply:
As stated above, we shouldn't be fetching this from Github at all. Rather, we should maintain a list of contexts that can be ignored in the Prow config.

Reply:
Yes, right now AFAIK all users are either:
- manually setting required jobs, in which case it shouldn't be a big deal to add them to a prow config entry
- autogenerating the required jobs from prow config with the branch protector, in which case we should reuse that.

I think this will have much less surprising behavior than attempting to sync with github behind a cache. Without some form of caching this will probably waste too many API calls.

Reply:
we should not :)

Reply:
Why can't a required context be inferred by the following combination in the config:
* job exists (thus can be controlled)
* job.always_run: true || job.run_if_changed != ""
* job.skip_report: false

Reply:
I disagree. having a list in Prow is going to be out of sync with the one in github (This is the issue we have with mungegithub right now). I am talking to the istio community to use the branch protection plugin, but this is going to take some time. Plus why should we force users to use the branch protection plugin ?

Reply:
Why would it be out of sync? making a job required on github (without the branch protector) is a manual action. It shouldn't be much more effort to add it to add a line with it to some config file.

Reply:
We could even make a tool that takes in a job name and updates the config file and makes the setting on github if necessary, but pulling it from github constantly is going to be wasteful and buggy.

Reply:
it will be out of sync, because that s problem that we have right now with mungegithub. We have multiple admins, and people doing changes out of our control. And you can tell people that they need to update it, but you cannot force them to.

Reply:
I tought of getting the required contexts from the PR using githubql as this would not take any additional api call, but then I saw you were already making an API call for combined status, so the same argument holds.

Reply:
@kargakis this only works for Prow jobs, what about other CI ?

Reply:
> I tought of getting the required contexts from the PR using githubql as this would not take any additional api call, but then I saw you were already making an API call for combined status, so the same argument holds.

No? Getting the required contexts does cost more.

Reply:
> it will be out of sync, because that s problem that we have right now with mungegithub. We have multiple admins, and people doing changes out of our control. And you can tell people that they need to update it, but you cannot force them to.

Not if you write it in a config in the first place, which also creates a public record and controls it via PR instead of admins randomly changing merge requirements silently. Nothing is out of sync if we push out the requirements to GitHub on config load instead of trying to keep up with GitHub.

> And you can tell people that they need to update it, but you cannot force them to.

Well I mean, you can't force them to use tide or the queue either, they can click merge manually even when it is deployed, but we don't optimize for this sort of broken usage either...

Reply:
That's fair, but it disable by default in your use case. So can we still get this in until we have a better approach ?